src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -1603,26 +1751,99 @@ private static void VerifyDerivedTypeConstraints(IEdmModel model, IEdmVocabulary
+
+                if (nextPiece[nextPiece.Length - 1] == ':')
+                {
+                    break;
+                }
+            }
+
+            string identifier = sb.ToString();
+
+            if (identifier != null && identifier.Length >= 2 && identifier[identifier.Length - 2] == ':')
+            {
+                anotherEscapeFunctionStarts = true;
+            }
+
+            bool isComposableRequired = identifier != null && identifier.Length >= 1 && identifier[identifier.Length - 1] == ':';
+
+            function = FindBestMatchForEscapeFunction(candidates, isComposableRequired, bindingType);
 
-            bool isComposableRequired = identifier.Length >= 2 && identifier[identifier.Length - 1] == ':';
-            IEdmFunction function = model.FindBoundOperations(bindingType)
-                .OfType<IEdmFunction>().FirstOrDefault(f => f.IsComposable == isComposableRequired && IsUrlEscapeFunction(model, f));
             if (function == null)
             {
+                // We need to throw because we have consumed segments from the queue and we don't put them back. This is fair because early checks did show an escape function bound to the type. 
                 throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_NoBoundEscapeFunctionSupported(bindingType.FullTypeName()));
             }
 
-            if (function.Parameters == null || function.Parameters.Count() != 2 || !function.Parameters.ElementAt(1).Type.IsString())
+            parenthesisExpression = function.Parameters.ElementAt(1).Name + "='" + (isComposableRequired ? identifier.Substring(0, identifier.Length - 1) : identifier) + "'";
+            qualifiedName = function.FullName();
+
+            return true;
+        }
+
+        private static IEdmFunction FindBestMatchForEscapeFunction(IEnumerable<IEdmFunction> candidtates, bool isComposable, IEdmType bindingType)
+        {
+            IEdmFunction bestCandidate = null;
+            foreach (IEdmFunction f in candidtates)
             {
-                throw ExceptionUtil.CreateBadRequestError(ODataErrorStrings.RequestUriProcessor_EscapeFunctionMustHaveOneStringParameter(function.FullName()));
+                // If the composability of the above found function does not match, we cannot use the function
+                if (f.IsComposable != isComposable)
+                {
+                    continue;
+                }
+
+                if (f.Parameters != null && f.Parameters.Count() == 2 && f.Parameters.ElementAt(1).Type.IsString())
+                {
+                    // The first parameter of the function is not an exact match for the binding type then try to find a better match.
+                    if (f.Parameters.FirstOrDefault().Type.Definition == bindingType)
+                    {
+                        return f;
+                    }
+                    else
+                    {
+                        bestCandidate = f;

```

Comment:
You are resetting the bestCandidate each time, which means that, if there is not an exact match, you'll pick the last one that matches (which, *I think*, is likely to be the base type, rather than the most derived type).  You could instead check to see if the current candidate has the current bestCandidate (if any) as a base type (recursively) and, if so, make the current candidate your new bestCandidate.